### ISSUE TITLE
Enhancement request for navigation: PgUp, PgDn + top, bottom

### ISSUE THREAD

**Issue by kseistrup** [4imothy/treegrep#8]
Would it be possible to support extra navigation so that `PgUp` and `PgDn` scrolls an entire page (perhaps an entire page minus one line, so as to maintain a sense of continuity) at a time? I often find myself with many matching lines and it can take a lot of scrolling, even with “big jump”, to skip many lines.

And perhaps also a set of “go to (top|bottom) of buffer” keys?

The `PgUp`/`PgDn` keys are unlikely to conflict with other keys, and they seem “intuitive” to me.

For top/bottom I was going to suggest `Home`/`End`, but some editors may use these keys for “(beginning|end) of line”, although that needn't be a stumbling block for treegrep.

At least `emacs` and `jed` are using `Esc+<` and `Esc+>` for top resp. bottom. I'm unsure about other editors.

**Comment by 4imothy** [comment]
Yes this is possible but requires some work on the menu code if I don't want to just scroll down $N$ times. The menu would have to implement a way to set the selected id and draw the matches around that.

I will implement this and thank you for all your suggestions but right now I am more focused on implementing a vim plugin similar to [fzf](https://github.com/junegunn/fzf).

**Comment by kseistrup** [comment]
Sure, please proceed at your own pace and in your own direction. We are not in a hurry.

PS: A thing to consider when scrolling a page at a time, is that window size may have changed since last time you made a query.

**Comment by 4imothy** [comment]
Got bored of vimscript and implemented this on [this branch](https://github.com/4imothy/treegrep/tree/redraw_around_selected_id) along with some other bindings to move up/down a path. I don't have easy access to a keyboard with the home/end or PageUp/PageDown keys though. Can you test to make sure the keys are being read correctly?

Should just be these commands
```
git clone https://github.com/4imothy/treegrep.git
cd treegrep
git checkout redraw_around_selected_id 
cargo run -- cmd -m   
```

Then press the keys and make sure they are working.

**Comment by kseistrup** [comment]
What a nice Sunday surprise! :pray: 

`Home`/`End` both work. :+1: 

`PgUp`/`PgDn` work in the sense that they move a screenful up or down.

However, if the cursor is at the top of the file and I press `PgDn`, then the cursor will be positioned on what was the last line of the initial screenful, _but_ that line will now be line number 11 counted from the bottom (i.e., there are 10 lines below it before the bottom of the screen). From now on the cursor will, no matter if I scroll a line or a page at a time, stay between top+10 lines (i.e. at line 11+) and bottom−10 lines. Only when the buffer's top line is actually at the top of the screen (or the buffer's bottom line is at the bottom) will the cursor move up beyond line 11 (or down beyond line 11 counted from the bottom).

This will take a while to get used to, I think. It's a good first attempt, though. :+1: 

What I had in mind was more like the movements of the pagers `more` / `less` / `most` / `moar` / `ov` (except that these don't have the concept of a cursor), that all move exactly one screenful at a time: after pressing `PgDn`, the line that was bottom+1 (i.e., just outside the bottom view) will now be line 1.

In numbers:

My terminal has 56 lines and one of these are used by the pager's status line, and so the pagers will initially all display line 1–55 + the status line. Scrolling down one screenful displays lines 56–110, another page down displays lines 111–165 and so on. And pressing `PgUp` will bring the previous screenful into the view.

And if the pager is displaying, say lines 111–165 and I use the cursor to scroll up one line, I will now see lines 110–164. Scrolling up one page from there will show me lines 55–109.

The pager [`moar`](https://github.com/walles/moar "Moar is a pager. It's designed to just do the right thing without any configuration.") has line numbers. Or you can “`cat -n alice[…].txt`” to obtain a file with line numbers, if you wish to see the movements yourself.

I think the bottom line of all this is (regardless of what I said about ”minus one line” in my original post) that I find it easier to navigate if `PgUp`/`PgDn` moves up/down exactly one screenful at a time. And if the cursor is at line 1 before I press `PgDn` it should be at the buffer's line screenful+1 and still at the top line (currently `tgrep` moves to the right line number of the buffer, but is now positioned at line 11 counted from the display bottom).

Generally there are two sets of movements:

1. There's the movement of the buffer relative to the dimensions of the terminal emulator — i.e., which line numbers are visible. That buffer can be scrolled 1 line, 1 $big_jump lines, or 1 screenful lines at a time.
2. There's the movement of the cursor:
   1. relative to the terminal emulator, and
   2. relative to the buffer lines.

_Unless_ the cursor will land ‘outside’ the visible page after a jump (i.e., line, big jump, or page), the buffer should remain fixed and only the cursor should move.

If the cursor _will_ land outside the current visible page, only the buffer should move the given number of lines, and the cursor should remain where it is relative to the terminal emulator.

I hope all this makes sense.

**Comment by kseistrup** [comment]
PS: No matter the cursor/buffer navigation, the correct line in the relevant is jumped to when selected in the buffer.

**Comment by 4imothy** [comment]
Yes I understand what you are saying I will try to implement that in the future. 

For now I will merge in the new branch as it has other good features as well, mainly the ability to jump to next/previous paths

**Comment by 4imothy** [comment]
I just pushed some changes in the [better_menu_controls](https://github.com/4imothy/treegrep/tree/better_menu_controls) branch. I only implemented the rule for the cursor not moving on paging but if you could check it out to make sure it is working for you and is what you described that would be great.

I will probably not do the rules for cursor only moving when displayed lines don't change on the other motions for a while and will now actually focus on creating the plugins I spoke about earlier.

**Comment by kseistrup** [comment]
It looks nice, and from the little testing I could do it looks like it does what I expect. Thanks.

**Comment by 4imothy** [comment]
With commit 7b9ba9670f20721f4a26a5216331a15c72fff46e, the goals of only the cursor moving when the target is in view and only the view moving when target is out of view is achieved. Let me know if anything is missing!
